api/filters/reference_instance_filter.go

```diff
@@ -0,0 +1,251 @@
+	if instance.ReferencedInstanceID == "" {
+		return next.Handle(req)
+	}
+
+	var marshal []byte
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	body := map[string]string{
+		instance_sharing.ReferencedInstanceIDKey: instance.ReferencedInstanceID,
+	}
+	marshal, err = json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+	return &web.Response{
+		Body:       marshal,
+		StatusCode: http.StatusOK,
+		Header:     headers,
+	}, nil
+}
+
+func (*referenceInstanceFilter) FilterMatchers() []web.FilterMatcher {
+	return []web.FilterMatcher{
+		{
+			Matchers: []web.Matcher{
+				web.Path(web.ServiceInstancesURL + "/**"),
+				web.Methods(http.MethodPost, http.MethodPatch, http.MethodGet),
+			},
+		},
+	}
+}
+
+func (rif *referenceInstanceFilter) handleProvision(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	servicePlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	isReferencePlan, err := rif.isReferencePlan(ctx, servicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	// Ownership validation
+	path := fmt.Sprintf("%s.%s", contextKey, rif.tenantIdentifier)
+	callerTenantID := gjson.GetBytes(req.Body, path).String()
+	if callerTenantID != "" {
+		err = rif.validateOwnership(req)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	parameters := gjson.GetBytes(req.Body, "parameters").Map()
+
+	referencedInstanceID, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	// epsilontal todo: should we validate that the input is string? can be any object for example...
+	if !exists {
+		return nil, util.HandleInstanceSharingError(util.ErrMissingReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+	req.Body, err = sjson.SetBytes(req.Body, instance_sharing.ReferencedInstanceIDKey, referencedInstanceID.String())
+	if err != nil {
+		return nil, err
+	}
+	_, err = rif.isReferencedShared(ctx, referencedInstanceID.String())
+	if err != nil {
+		return nil, err
+	}
+	log.C(ctx).Infof("Reference Instance Provision passed successfully. Called by \"%s\", instanceID: \"%s\"", callerTenantID, referencedInstanceID)
+	return next.Handle(req)
+}
+
+func (rif *referenceInstanceFilter) handleServiceUpdate(req *web.Request, next web.Handler) (*web.Response, error) {
+	// we don't want to allow plan_id and/or parameters changes on a reference service instance
+	resourceID := req.PathParams["resource_id"]
+	ctx := req.Context()
+
+	dbInstanceObject, err := rif.getObjectByID(ctx, types.ServiceInstanceType, resourceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	if instance.ReferencedInstanceID == "" {
+		return next.Handle(req)
+	}
+
+	_, err = rif.isValidPatchRequest(req, instance)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("%s.%s", contextKey, rif.tenantIdentifier)
+	callerTenantID := gjson.GetBytes(req.Body, path).String()
```

Comment:
if the tenant is not relevant for this flow why do you look it up? we don't need irrelevant info in the log..